### PR TITLE
perf: Avoid string copying during JSON parsing

### PR DESCRIPTION
**Pull request by huskar-t**
This submission implements a performance optimization for JSON parsing by avoiding the copying of strings into byte arrays.

### CODE REVIEW

**Comment by ngjaying** [comment]
Could you please also paste your benchmark result for the hack package.

**Comment by huskar-t** [comment]
> Could you please also paste your benchmark result for the hack package.

image

**Comment by ngjaying** [comment]
> > Could you please also paste your benchmark result for the hack package.
> 
> image

Amazing improvement! Are there any risks? Additionally, is it better to rename the package 'hack'? How do you like to move it to cast package, which already have functions of type conversions.

**Comment by huskar-t** [comment]
@ngjaying 
Since Go strings are immutable, the byte arrays must be immutable.
```go
func main() {
	str := "abc"
	a := string2bytes(str)
	fmt.Println("len:", len(a))
	fmt.Println("cap:", cap(a))
	fmt.Println(a)
	a[0] = 'd'
	fmt.Println(a)
	fmt.Println(str)
}

func string2bytes(str string) []byte {
	return unsafe.Slice(unsafe.StringData(str), len(str))
}
// output:
// len: 3
// cap: 3
// [97 98 99]

// segment fault at a[0] = 'd'
```
it can be used if the byte array is read-only

**Comment by huskar-t** [comment]
@ngjaying 
Because len=cap, the append operation will expand the byte array, so the append operation is safe.
```go
func main() {
	str := "abc"
	a := string2bytes(str)
	fmt.Println("len:", len(a))
	fmt.Println("cap:", cap(a))
	fmt.Println(a)
	fmt.Printf("address of bytes before append : %x\n", (*reflect.SliceHeader)(unsafe.Pointer(&a)).Data)
	a = append(a, 'e')
	fmt.Printf("address of bytes after  append : %x\n", (*reflect.SliceHeader)(unsafe.Pointer(&a)).Data)
	a[0] = 'd'
	fmt.Printf("address of bytes after  change : %x\n", (*reflect.SliceHeader)(unsafe.Pointer(&a)).Data)
	fmt.Println(len(a))
	fmt.Println(cap(a))
	fmt.Println(a)
	fmt.Println(str)
}

func string2bytes(str string) []byte {
	return unsafe.Slice(unsafe.StringData(str), len(str))
}

//output
//len: 3
//cap: 3
//[97 98 99]
//address of bytes before append : ca7ea5
//address of bytes after  append : c00001c0f8
//address of bytes after  change : c00001c0f8
//4
//8
//[100 98 99 101]
//abc
```

File: internal/hack/hack.go [hunk 1]
```diff
@@ -0,0 +1,7 @@
+package hack
```
Comment (ngjaying) [inline internal/hack/hack.go:1]: Suggest to move this to pkg/cast package.

**Review by ngjaying — CHANGES_REQUESTED** [review]

File: internal/hack/hack.go (hunk 1 above)
Comment (huskar-t) [reply internal/hack/hack.go]: done

**Review by ngjaying — APPROVED** [review]

**Comment by ngjaying** [comment]
Thank you!

**Pull request merged**